odoo/http.py

```diff
@@ -415,6 +368,133 @@ def serialize_exception(exception):
         'context': getattr(exception, 'context', {}),
     }
 
+
+# =========================================================
+# File Streaming
+# =========================================================
+
+KEEP = object()  # sentinel
+
+@dataclass
+class StreamData:
+    type = ''  # 'data' or 'path' or 'url'
+    data = None
+    path = None
+    url = None
+
+    mimetype = None
+    as_attachment = False
+    download_name = None
+    conditional = True
+    etag = True
+    last_modified = None
+    max_age = None
+    size = None
+
+    @classmethod
+    def from_path(cls, path):
+        path = file_path(path)
+        return cls(
+            type='path',
+            path=path,
+            as_attachment=as_attachment,
+            size=os.stat(path).size
+        )
+
+    @classmethod
+    def from_attachment(cls, attachment):
+        self = cls(
+            mimetype=attachment.mimetype,
+            as_attachment=True,
+            download_name=attachment.name,
+            conditional=True,
+            etag=attachment.checksum,
+            max_age=STATIC_CACHE_LONG,
+        )
+
+        if attachment.stored_fname:
+            stat = os.stat
+            self.type = 'path'
+            self.path = werkzeug.security.safe_join(
+                os.path.abspath(config.filestore()),
+                attachment.stored_fname
+            )
+            self.last_modified = stat.mtime
+            self.size = stat.size
+
+        elif attachment.db_datas:
+            self.type = 'data'
+            self.data = attachment.raw
+            self.size = len(self.data)
+
+        elif attachment.url:
+            self.type = 'url'
+            self.url = attachment.url
+
+            # When the URL targets a file located in an addon, assume it
+            # is a path instead. It saves a 3xx response.
+            if (attachment.url or '').startswith('/'):
```

Comment:
The heuristic is broken by most addon files not normally being served e.g. you can't just go and dereference python files in the module in order to look at that's in there. Only the stuff that's in `/<module>/static` is "public".

Reply:
Changed in favor of:

```py
if (attachment.url or '')[1:].partition('/')[0] in root.statics:
```

[1:] is removeprefix('/')